nntrainer/dataset/data_producers.h

```diff
@@ -43,12 +44,39 @@ class DataProducer {
 
   /**
    * @brief create an iteration
+   * @todo rename this to BatchGenerator
    * @return Iteration iteration, if std::get<0>(retval) == true means end of
    * iteration, at the end of the iteration, it's responsibility of @a this to
    * shuffle.
    */
   using Generator = std::function<Iteration(void)>;
 
+  /**
+   * @brief Sample represents a view of single element which can be fed to the
+   * model. It is the smallest unit to produce a data
+   * @return std::get<0>(Sample) denotes inputs
+   * @return std::get<1>(Sample) denotes labels
+   */
+  using Sample = std::tuple<std::vector<Tensor *>, std::vector<Tensor *>>;
+
+  /**
+   * @brief generator callable type which will fill a sample
+   * @todo rename this to Generator.
+   * @param[in] index current index with range of [0, size() - 1]. If
+   * size() == SIZE_UNDEFINED, this parameter can be ignored
+   * @param[out] inputs allocate tensor before expected to be filled by this
+   * function
+   * @param[out] labels allocate tensor before expected to be filled by this
+   * function function.
+   * @return bool true if this is the last sample, samples will NOT be ignored
+   * and should be used, or passed at will of caller
+   *
+   */
+  using Generator_sample =
+    std::function<bool(unsigned int, /** index */
+                       std::vector<Tensor *> & /** inputs */,
```

Comment:
const vector?

Reply:
This is changing so it's not const

Reply:
I may be wrong, but in my understanding, as the contents in the tensor are changing but not the tensor itself, the vector can be a const.